### PR TITLE
CFE-2572: Add sys.policy_entry variables

### PR DESCRIPTION
**Pull request by vpodzime**
This is adding 3 new variables to the "sys" scope:
  sys.policy_entry_filename - full absolute path of policy file (inclufing .cf or other extension)
  sys.policy_entry_dirname  - full absolute path of directory
  sys.policy_entry_basename - only basename (file name, including .cf or other extension)

Changelog: Title

### CODE REVIEW

**Comment by nickanderson** [comment]
Seems OK to me. Exactly what I wanted from CFE-2572.

File: libpromises/generic_agent.c [hunk 1]
```diff
@@ -446,6 +446,20 @@ void LoadAugments(EvalContext *ctx, GenericAgentConfig *config)
...
+void AddPolicyEntryVariables (EvalContext *ctx, GenericAgentConfig *config)
+{
+    char *abs_input_path = GetAbsolutePath(config->input_file);
+    /* both dirname() and basename() may actually modify the string they are given (see man:basename(3)) */
+    char *dirname_path = xstrdup(abs_input_path);
+    char *basename_path = xstrdup(abs_input_path);
+    EvalContextVariablePutSpecial(ctx, SPECIAL_SCOPE_SYS, "policy_entry_filename", abs_input_path, CF_DATA_TYPE_STRING, "source=agent");
+    EvalContextVariablePutSpecial(ctx, SPECIAL_SCOPE_SYS, "policy_entry_dirname", dirname(dirname_path), CF_DATA_TYPE_STRING, "source=agent");
+    EvalContextVariablePutSpecial(ctx, SPECIAL_SCOPE_SYS, "policy_entry_basename", basename(basename_path), CF_DATA_TYPE_STRING, "source=agent");
+    free (abs_input_path);
```
Comment (jimis) [inline libpromises/generic_agent.c]: remove the space before parenthesis, and try to wrap at 78. See HACKING.md.

File: libpromises/generic_agent.c [hunk 2]
```diff
@@ -446,6 +446,20 @@ void LoadAugments(EvalContext *ctx, GenericAgentConfig *config)
     free(def_json);
 }
 
+void AddPolicyEntryVariables (EvalContext *ctx, GenericAgentConfig *config)
```
Comment (jimis) [inline libpromises/generic_agent.c]: constify parameters that are not modified

File: libpromises/files_names.c [hunk 3]
```diff
@@ -589,6 +623,35 @@ bool CompressPath(char *dest, size_t dest_size, const char *src)
 
 /*********************************************************************/
 
+char *GetAbsolutePath(const char *path)
+/**
+ * WARNING: Remember to free return value.
+ **/
```
Comment (jimis) [inline libpromises/files_names.c]: Function comments go in doxygen-like style before, not after the function title. You may also want to be a bit more verbose about what this function does. 

Also applies to the other function.

File: libpromises/files_names.c [hunk 4]
```diff
@@ -589,6 +623,35 @@ bool CompressPath(char *dest, size_t dest_size, const char *src)
...
+    if (IsAbsoluteFileName(path))
+    {
+        CompressPath(abs_path, PATH_MAX, path);
+        return xstrdup(abs_path);
+    }
+    else
+    {
+        /* the full_path can potentially be long (with many '../' parts)*/
+        char full_path[2 * PATH_MAX] = { 0 };
+        if (!getcwd(full_path, PATH_MAX))
```
Comment (jimis) [inline libpromises/files_names.c]: Explicit comparison with NULL pointer. Use the `!` operator only for boolean comparisons.

File: libpromises/files_names.c [hunk 5]
```diff
@@ -170,6 +170,40 @@ char *JoinSuffix(char *path, size_t path_size, const char *leaf)
...
+
+    if (path_len + len + 1 > path_size)
+    {
+        Log(LOG_LEVEL_ERR, "JoinPaths: Internal limit reached. Tried to add %s to %s",
+              leaf_path, path);
+        return NULL;
+    }
+
+    /* make sure there's a FILE_SEPARATOR between path and leaf_path */
+    if (!IsFileSep(path[path_len - 1]) && !IsFileSep(leaf_path[0]))
```
Comment (jimis) [inline libpromises/files_names.c]: what if `path_len` is zero

File: libpromises/files_names.c [hunk 6]
```diff
@@ -170,6 +170,40 @@ char *JoinSuffix(char *path, size_t path_size, const char *leaf)
...
+Just like the JoinSuffix() above, but makes sure there's a FILE_SEPARATOR
+between @path and @leaf_path.
+*/
+{
+    size_t len = strlen(leaf_path);
+    size_t path_len = strlen(path);
+
+    if (Chop(path, path_size) == -1)
+    {
+        Log(LOG_LEVEL_ERR, "Chop was called on a string that seemed to have no terminator");
```
Comment (jimis) [inline libpromises/files_names.c]: return NULL?

File: libpromises/files_names.c [hunk 7]
```diff
@@ -170,6 +170,40 @@ char *JoinSuffix(char *path, size_t path_size, const char *leaf)
...
+char *JoinPaths(char *path, size_t path_size, const char *leaf_path)
+/*
+Just like the JoinSuffix() above, but makes sure there's a FILE_SEPARATOR
+between @path and @leaf_path.
+*/
+{
+    size_t len = strlen(leaf_path);
+    size_t path_len = strlen(path);
+
+    if (Chop(path, path_size) == -1)
```
Comment (jimis) [inline libpromises/files_names.c]: Shouldn't this be `path_size-1` according to the comments in `Chop()`?

Comment (vpodzime) [reply libpromises/files_names.c]: Yeah, I believe it should be ``path_size-1``. I didn't check the code I copied over from ``JoinSuffix()``.

File: libpromises/files_names.c (hunk 6 above)
Comment (vpodzime) [reply libpromises/files_names.c]: ditto

File: libpromises/files_names.c (hunk 5 above)
Comment (vpodzime) [reply libpromises/files_names.c]: Good catch! I'll also add tests for this case.

File: libpromises/files_names.c (hunk 3 above)
Comment (vpodzime) [reply libpromises/files_names.c]: Done.

File: libpromises/files_names.c (hunk 4 above)
Comment (vpodzime) [reply libpromises/files_names.c]: Done.

File: libpromises/generic_agent.c (hunk 2 above)
Comment (vpodzime) [reply libpromises/generic_agent.c]: Done. I also made the function ``static``.

File: libpromises/generic_agent.c (hunk 1 above)
Comment (vpodzime) [reply libpromises/generic_agent.c]: Done.

**Comment by olehermanse** [comment]
+1

Looks very good overall, I didn't review in detail, because jimis already has.

**Comment by nickanderson** [comment]
docs: 
https://github.com/cfengine/documentation/pull/1710

**Comment by vpodzime** [comment]
> +1 looks good. @vpodzime can you submit a jenkins build and link it here?

https://ci.cfengine.com/job/testing-enterprise-pr/541

**Comment by jimis** [comment]
All green, merging, thanks @vpodzime !

**Pull request merged**